agent/consul/state/kvs.go

```diff
@@ -148,12 +149,17 @@ func (s *Store) kvsSetTxn(tx *memdb.Txn, idx uint64, entry *structs.DirEntry, up
 	// session for a new entry is "no session".
 	if !updateSession {
 		if existing != nil {
-			entry.Session = existing.(*structs.DirEntry).Session
+			entry.Session = existing.Session
 		} else {
 			entry.Session = ""
 		}
 	}
 
+	// skip write if the entry did not change
+	if existing != nil && existing.Equal(entry) {
+		return nil
```

Comment:
@ShimmerGlass  Hi! I'm sorry for writing it here and after 11 months since your commits, but it looks like it is necessary to preserve `ModifyIndex` for unchanged entries before quit with 'nil':
```
if existing != nil && existing.Equal(entry) {
    entry.ModifyIndex = existing.ModifyIndex
    return nil
}
```

That ModifyIndex is returned as a result of transactional changes of KVStore (through /v1/txn) and people might relay on that result for future CAS-changes. And without preserving it as it was, people will bump into the error "failed to set key %q, index is stale" (as I was)